### PR TITLE
FreeRTOS+TCP: Let xDataLength always mean "total number of bytes"

### PR DESCRIPTION
**Pull request by htibosch**
Description
-----------
A network buffer (`NetworkBufferDescriptor_t`) has a field `xDataLength`. This is the total length of the buffer. It includes the 14-byte Ethernet header and everything after that.

Just recently, @aggarg  wanted to add a length check in the function `vARPGenerateRequestPacket()`, something like:

~~~c
	configASSERT( pxNetworkBuffer->xDataLength >= sizeof(ARPPacket_t) )
~~~

This function failed because there are two cases in which `xDataLength` has a different meaning.

- In an outgoing ICMP/ping packet, it means the number of bytes to send plus the size of an ICMP header.
- In an outgoing UDP packet, it is the payload size ( without the UDP header ).

In order to avoid confusion, I'd like to correct that: `xDataLength` will always contain the total number of bytes.

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by htibosch** [comment]
When UDP packets are sent or received, the field `xDataLength` still contained the size of the nett payload. I believe that this was the ultimate situation in which `xDataLength` referred to a payload size.

A simple change in `FreeRTOS_Sockets.c` in `FreeRTOS_recvfrom()`:

~~~c
-	lReturn = ( int32_t ) pxNetworkBuffer->xDataLength;
+	lReturn = ( int32_t ) pxNetworkBuffer->xDataLength - sizeof( UDPPacket_t );
~~~

In `FreeRTOS_sendto()` there is this change:

~~~c
-	pxNetworkBuffer->xDataLength = xTotalDataLength;
+	pxNetworkBuffer->xDataLength = xTotalDataLength + sizeof( UDPPacket_t );
~~~

In `FreeRTOS_DNS.c`, added a variable `uxPayloadSize` in a few occasions:

~~~c
-	if( pxNetworkBuffer->xDataLength >= sizeof( DNSMessage_t ) )
+	size_t uxPayloadSize = pxNetworkBuffer->xDataLength - sizeof( UDPPacket_t );
+	if( uxPayloadSize >= sizeof( DNSMessage_t ) )
~~~

In `FreeRTOS_IP.c`, there is an extra sanity check on the UDP field `usLength`:

	if ( ... && FreeRTOS_ntohs( pxUDPPacket->xUDPHeader.usLength ) >= sizeof( UDPHeader_t ) )

All tested again with and without: DNS, LLMNR, NBNS, and UDP messages in both directions.
I also check the values of variables by stepping through the code.

Hein

File: libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_Sockets.c [hunk 1]
```diff
@@ -704,7 +704,7 @@ EventBits_t xEventBits = ( EventBits_t ) 0;
 
 		/* The returned value is the data length, which may have been capped to
 		the receive buffer size. */
-		lReturn = ( int32_t ) pxNetworkBuffer->xDataLength;
+		lReturn = ( int32_t ) pxNetworkBuffer->xDataLength - sizeof( UDPPacket_t );
```
[the remark this answers is not in the document]
Comment (htibosch) [reply libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_Sockets.c]: > Hasn't the header already been removed before this function was called?

Indeed, the headers were not removed, the packet still contains them when `FreeRTOS_recvfrom()` is called.
Currently this function would look at a full packet ( >= 42 bytes ), where the `xDataLength` field would reflect the length of the payload.

A little further the function is accessing the buffer as:

~~~c
    &( pxNetworkBuffer->pucEthernetBuffer[ ipUDP_PAYLOAD_OFFSET_IPv4 ] )
~~~

Note that `ipUDP_PAYLOAD_OFFSET_IPv4` == `sizeof( UDPPacket_t )` == 42.

The aim of this PR is to let `xDataLength` reflect the length of the buffer, independent from the context.

File: libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_DNS.c [hunk 2]
```diff
@@ -1326,14 +1326,10 @@ BaseType_t xDoStore = xExpected;
...
-										 sizeof( EthernetHeader_t ) + sizeof( IPHeader_t );
 
-					/* The field xDataLength was set to the length of the UDP payload.
-					The answer (reply) will be longer than the request, so the packet
-					must be duplicated into a bigger buffer */
-					pxNetworkBuffer->xDataLength = xDataLength;
-					pxNewBuffer = pxDuplicateNetworkBufferWithDescriptor( pxNetworkBuffer, xDataLength + 16 );
+					/* The field xDataLength was set to the total length of the UDP packet,
+					i.e. the payload size plus sizeof( UDPPacket_t ). */
+					pxNewBuffer = pxDuplicateNetworkBufferWithDescriptor( pxNetworkBuffer, pxNetworkBuffer->xDataLength + 16 );
```
[the remark this answers is not in the document]
Comment (htibosch) [reply libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_DNS.c]: The `+16` was there already, in two cases.
In the first case it means `sizeof( LLMNRAnswer_t )`,
in the second case it means `sizeof( NBNSAnswer_t )`.
Now adapted.

**Comment by htibosch** [comment]
Maybe some clarification about receiving UDP packets:

NetworkInterface receives a UDP packet :

	42 bytes headers
	10 bytes payload
	18 bytes padding
	- - - - - - - - - - - -
	60 bytes = xDataLength

The packet is sent to a queue and waits for the IP-task.

IP-task receives the network buffer and processes it in `prvProcessIPPacket()`:

	It corrects `xDataLength`: the padding is discarted:

	42 bytes headers
	10 bytes payload
	- - - - - - - - - - - -
	52 bytes = xDataLength

API call `FreeRTOS_recvfrom()`:

	Copy method: it copies the 10-byte payload to a user buffer.
	Zero-copy: it returns a pointer to the Network Buffer at the offset of the payload.

The user receives an object of type `void *` which can be read from ( character bytes ), and which shall be released by calling :

~~~c
    void FreeRTOS_ReleaseUDPPayloadBuffer( void *pvBuffer )
~~~

We don't know in advance if ``xDataLength` is correct. Therefor we take the smaller of the 2 calculated payloads, with a minimum of 0 bytes.
( not sure if a zero-length UDP-packet has any use in real life ).
Hein

File: libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_IP.c [hunk 3]
```diff
@@ -1611,23 +1612,24 @@ uint8_t ucProtocol;
...
+						 * the payload size plus the UDP header ( 8 bytes ).
+						 * Set `xDataLength` to the size of the headers,
+						 * plus the lower of the two calculated payload sizes.
 						 */
 
-						pxNetworkBuffer->xDataLength -= sizeof( UDPPacket_t );
-						if( ( FreeRTOS_ntohs( pxUDPPacket->xUDPHeader.usLength ) - sizeof( UDPHeader_t ) ) <
-								pxNetworkBuffer->xDataLength )
+						uxPayloadSize_1 = pxNetworkBuffer->xDataLength - sizeof( UDPPacket_t );
+						uxPayloadSize_2 = FreeRTOS_ntohs( pxUDPPacket->xUDPHeader.usLength ) - sizeof( UDPHeader_t );
```
[the remark this answers is not in the document]
Comment (htibosch) [reply libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_IP.c]: > The first line above uses sizeof( UDPPacket_t ) while the second uses
> sizeof( UDPHeader_t ). Are they not the same size?

Within the library, the word "Packet" refers to an Ethernet packet, including the layers 2, 3, and 4.
A `UDPHeader_t` and `TCPHeader_t` are just the headers used in the Transport layer ( 4 ).

In `prvProcessIPPacket()`, the UDP payload is calculated twice:

~~~c
    uxPayloadSize_1 = pxNetworkBuffer->xDataLength - sizeof( UDPPacket_t );
    uxPayloadSize_2 = FreeRTOS_ntohs( pxUDPPacket->xUDPHeader.usLength ) - sizeof( UDPHeader_t );
~~~

As you know, the length field `usLength` of the UDP header includes the size of the header itself.
`xDataLength` includes the whole packet, plus the payload bytes.

[the remark this answers is not in the document]
Comment (dan4thewin) [reply libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_IP.c]: I understand.  Thank you.

**Comment by yuhui-zheng** [comment]
Since we got two approvals and tests are passing, I'm merging this PR as part of the oncall duty.

**Pull request merged**